### PR TITLE
test: fork choice reorg after newly justified slot

### LINKED ISSUE #114: tests: add tests for complex justification boundaries
ok @unnawut for now I've this test that is failing. I guess something is strange right now in the justification boundary check in fork choice head selection. We should open an issue probably to fix this. For now I'll completely remove this test to avoid having a mess of fixes in this PR and merge asap, we can fix the rest in a followup PR.

```python
def test_reorg_with_justification_boundary(
    fork_choice_test: ForkChoiceTestFiller,
) -> None:
    """
    Fork choice should prefer justifiable forks over non-justifiable ones.

    Scenario
    --------
    Two forks compete where one extends past justifiable boundaries:
    - Fork A: extends to slot 7 (NOT justifiable - not ≤5, not perfect square, not pronic)
    - Fork B: extends to slot 5 (justifiable - within ≤5 threshold)

    Justification Rules (from finalized slot 0):
    - Justifiable: delta ≤ 5, perfect squares (9, 16, 25...), pronic (6, 12, 20...)
    - NOT Justifiable: 7, 8, 10, 11, 13, 14, 15, etc.

    Expected Behavior
    -----------------
    1. Fork A extends to slot 7 (longer chain, 7 blocks)
    2. Fork B extends to slot 5 (shorter chain, 5 blocks)
    3. Fork B should win despite being shorter because:
       - Fork B's head (slot 5) is justifiable (5 ≤ 5)
       - Fork A's head (slot 7) is NOT justifiable
       - Fork choice should prefer justifiable heads for safety

    Why This Matters
    ----------------
    Justification boundaries are a critical safety mechanism:
    - Prevents the head from extending indefinitely past justifiable slots
    - Ensures the canonical chain stays within attestable range
    - Protects against long-range attacks
    - Maintains finality guarantees
    """
    fork_choice_test(
        steps=[
            # Common base at slot 1 (both forks will diverge from here)
            BlockStep(
                block=BlockSpec(slot=Slot(1), label="common"),
                checks=StoreChecks(head_slot=Slot(1), head_root_label="common"),
            ),
            # Fork A extends all the way to slot 7
            BlockStep(
                block=BlockSpec(slot=Slot(2), parent_label="common", label="fork_a_2"),
                checks=StoreChecks(head_slot=Slot(2), head_root_label="fork_a_2"),
            ),
            BlockStep(
                block=BlockSpec(slot=Slot(3), parent_label="fork_a_2", label="fork_a_3"),
                checks=StoreChecks(head_slot=Slot(3), head_root_label="fork_a_3"),
            ),
            BlockStep(
                block=BlockSpec(slot=Slot(4), parent_label="fork_a_3", label="fork_a_4"),
                checks=StoreChecks(head_slot=Slot(4), head_root_label="fork_a_4"),
            ),
            BlockStep(
                block=BlockSpec(slot=Slot(5), parent_label="fork_a_4", label="fork_a_5"),
                checks=StoreChecks(head_slot=Slot(5), head_root_label="fork_a_5"),
            ),
            BlockStep(
                block=BlockSpec(slot=Slot(6), parent_label="fork_a_5", label="fork_a_6"),
                checks=StoreChecks(head_slot=Slot(6), head_root_label="fork_a_6"),
            ),
            # Fork A at slot 7 (NOT justifiable from slot 0: not ≤5, not square, not pronic)
            BlockStep(
                block=BlockSpec(slot=Slot(7), parent_label="fork_a_6", label="fork_a_7"),
                checks=StoreChecks(
                    head_slot=Slot(7),
                    head_root_label="fork_a_7",  # Fork A leads (7 blocks total)
                ),
            ),
            # Fork B competes, building from common ancestor
            BlockStep(
                block=BlockSpec(slot=Slot(2), parent_label="common", label="fork_b_2"),
                checks=StoreChecks(head_slot=Slot(7), head_root_label="fork_a_7"),
            ),
            BlockStep(
                block=BlockSpec(slot=Slot(3), parent_label="fork_b_2", label="fork_b_3"),
                checks=StoreChecks(head_slot=Slot(7), head_root_label="fork_a_7"),
            ),
            BlockStep(
                block=BlockSpec(slot=Slot(4), parent_label="fork_b_3", label="fork_b_4"),
                checks=StoreChecks(head_slot=Slot(7), head_root_label="fork_a_7"),
            ),
            # Fork B stops at slot 5 (justifiable: 5 ≤ 5)
            # Fork B should win despite being shorter because:
            # - Fork B's head (slot 5) is justifiable
            # - Fork A's head (slot 7) is NOT justifiable
            # - Fork choice should prefer justifiable heads
            BlockStep(
                block=BlockSpec(slot=Slot(5), parent_label="fork_b_4", label="fork_b_5"),
                checks=StoreChecks(
                    head_slot=Slot(5),
                    head_root_label="fork_b_5",  # Fork B wins (justifiable)
                ),
            ),
        ],
    )
```

_Originally posted by @tcoratger in https://github.com/leanEthereum/leanSpec/pull/105#discussion_r2486037106_

### PR DESCRIPTION
**Pull request by unnawut**
## 🗒️ Description

After a bit of digging I realize that the fork choice tests:

1. Have only one attestation per block (by the proposer itself) so they never justified beyond genesis
2. Never needed to check for justification slot before, only checking heaviest chain back to genesis, so they have always worked without new justifications
3. Head determination does not need to be aware of the actual justification rule. It only needs to know what is the latest justified slot and update head based on that. So checking for justification rule doesn't belong in reorg tests, any triggered justification is suffice

So in order to test reorg based on new justifications:

- Added `SignedAttestationSpec` and `_build_signed_attestation_from_spec()` to generate attestations as part of `BlockSpec`
- Added more attestations (at least 2 attesters in addition to the proposer) to trigger a new justification

Testing justification rules like perfect square and pronic number will be added to [test_attestation_target_selection.py](https://github.com/leanEthereum/leanSpec/blob/main/tests/consensus/devnet/fc/test_attestation_target_selection.py) instead of this reorg test module -> Already done. Submitting as the next PR

Closes #114

## 🔗 Related Issues or PRs

#114, #105

## ✅ Checklist

- [x] Ran `tox` checks to avoid unnecessary CI fails:
    ```console
    uvx tox
    ```
- [x] Considered adding appropriate tests for the changes.
- [x] Considered updating the online docs in the [./docs/](/leanEthereum/leanSpec/tree/main/docs/) directory.

### CODE REVIEW

File: packages/testing/src/consensus_testing/test_fixtures/fork_choice.py [hunk 1]
```diff
@@ -256,8 +267,27 @@ def _build_block_from_spec(
...
+                    # Extract the Attestation message and signature
+                    attestations.append(signed_attestation.message)
+                    attestation_signatures.append(signed_attestation.signature)
+                else:
+                    # Already a SignedAttestation, extract the message
+                    attestations.append(attestation.message)
+                    attestation_signatures.append(attestation.signature)
+
+        # Build block with collected attestations
+        body = BlockBody(attestations=Attestations(data=attestations))
```
Comment (tcoratger) [inline packages/testing/src/consensus_testing/test_fixtures/fork_choice.py:290]: Ooooh, thanks for this! I completely forgot to check that. The tests worked fine but didn't really mean much before then. Thanks for pointing this out!
